src/librustc_typeck/check/pat.rs

```diff
@@ -371,17 +378,71 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             self.emit_err_pat_range(
                 span, begin.span, end.span, lhs_fail, rhs_fail, lhs_ty, rhs_ty
             );
```

Comment:
You'll have to change `emit_err_pat_range` to return the error instead of emitting it.

All of this then becomes 

```
            return self.emit_err_pat_range(
                span, begin.span, end.span, lhs_fail, rhs_fail, lhs_ty, rhs_ty
                span, begin.span, end.span, lhs_fail, rhs_fail, lhs_ty, rhs_ty
            );
```

which lets you have the `else` clause implicit, removing one indentation level.